toolchain/base/value_store_chunk.h

```diff
@@ -0,0 +1,172 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_VALUE_STORE_CHUNK_H_
+#define CARBON_TOOLCHAIN_BASE_VALUE_STORE_CHUNK_H_
+
+#include <bit>
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "common/check.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/MemAlloc.h"
+#include "toolchain/base/mem_usage.h"
+
+namespace Carbon::Internal {
+
+// Ids which are stored in a `ValueStore` have a `ValueType` which indicates the
+// type of value held in the `ValueStore`.
+template <class IdT>
+concept IdHasValueType = requires { typename IdT::ValueType; };
+
+// The max size of each chunk allocation for `ValueStore`. This is based on TLB
+// page sizes for the target platform.
+//
+// See https://docs.kernel.org/admin-guide/mm/hugetlbpage.html
+//
+// A 4K chunk size outperforms a 1M chunk size on Linux-x64 and MacOS-arm64 in
+// benchmarks and when running file_test.
+//
+// TODO: The previous SmallVector<ValueType> seems to outperform 4K chunks (they
+// may be slower by up to 5%) in benchmarks. Find ways to make chunking faster.
+// Should successive chunks get larger in size? That will greatly complicate
+// math for choosing a chunk though.
+template <class IdT>
+  requires(IdHasValueType<IdT>)
+static constexpr auto PlatformChunkMaxAllocationBytes() -> int32_t {
+#if !defined(NDEBUG) || LLVM_ADDRESS_SANITIZER_BUILD
+  // Use a small size in unoptimized builds to ensure multiple chunks get used.
+  // And do the same in ASAN builds to reduce bookkeeping overheads. Using large
+  // allocations (e.g. 1M+) incurs a 10x runtime cost for our tests under ASAN.
+  return sizeof(typename IdT::ValueType) * 5;
+#else
+  // TODO: Should ia64 use 1M or 4M? Should Windows and Mac use different sizes?
+
+  // x64 CPUs support 4K and 2M page sizes, but we see 1M is slower than 1K with
+  // tcmalloc in opt builds for our tests.
+  //
+  // FIXME: More benchmarking needed.
+  return 4 * 1024;
+#endif
+}
+
+// The number of elements stored in each chunk allocation.
+//
+// The number must be a power of two so that that there are no unused values in
+// bits indexing into the allocation.
+template <class IdT>
+  requires(IdHasValueType<IdT>)
+constexpr auto PlatformChunkCapacity() -> int32_t {
+  constexpr auto MaxElements =
+      PlatformChunkMaxAllocationBytes<IdT>() / sizeof(typename IdT::ValueType);
+  return std::bit_floor(MaxElements);
+}
+
+// The number of bits needed to index each element in a chunk allocation.
+template <class IdT>
+  requires(IdHasValueType<IdT>)
+constexpr auto PlatformChunkIndexBits() -> int32_t {
+  static_assert(PlatformChunkCapacity<IdT>() > 0);
+  return std::bit_width(uint32_t{PlatformChunkCapacity<IdT>() - 1});
```

Comment:
I think we typically use `static_cast`, e.g.:
```suggestion
  return std::bit_width(static_cast<uint32_t>(PlatformChunkCapacity<IdT>() - 1));
```

Reply:
The style guide prefers this syntax when possible, for primitives specifically. It does bounds checking for you at compile time. It doesn't compile if it can't.